prometheus metrics support

dfengshi opened the pull request:
attempt at addressing https://github.com/banzaicloud/bank-vaults/issues/72 similar to the banzai-charts helm chart

@bonifaido Hello, as we discussed in the issue #72 , here's a quick attempt at addressing the issue with a sidecar statsd container.

Please feel free to give copious feedback/help as I'm new to the operator pattern and go in general. Apologies if there are any glaring issues, I'll do my best to address. I do have a few questions about the volume mount and the TLS Env.

Thanks!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/banzaicloud/bank-vaults?pullRequest=125) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/banzaicloud/bank-vaults?pullRequest=125) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/banzaicloud/bank-vaults?pullRequest=125) it.</sub>

bonifaido commented:
The TLS volumes are needed since the Vault Api is exposed through self signed certificate, so the CA is needed.

dfengshi commented:
@bonifaido thanks, my bad

bonifaido commented:
Only the formatting is now what needs to be fixed (go fmt)

bonifaido commented on operator/pkg/stub/handler.go:
```diff
@@ -417,6 +417,28 @@ func statefulSetForVault(v *v1alpha1.Vault) (*appsv1.StatefulSet, error) {
+								ContainerPort: 9102,
+								Protocol:      "TCP",
+							}},
+							VolumeMounts: withStatsDVolumeMount(v, withCredentialsVolumeMount(v, []v1.VolumeMount{})),
```
`withCredentialsVolumeMount` is not needed for statsd. It is used to access on-cloud resources such as S3 or KMS for example.

bonifaido commented on operator/pkg/stub/handler.go:
```diff
@@ -484,6 +506,14 @@ func withTLSVolumeMount(v *v1alpha1.Vault, volumeMounts []v1.VolumeMount) []v1.V
 	return volumeMounts
 }
 
+func withStatsDVolumeMount(v *v1alpha1.Vault, volumeMounts []v1.VolumeMount) []v1.VolumeMount {
```
The `statsd-mapping` volume has to be created as in: https://github.com/banzaicloud/banzai-charts/blob/master/vault/templates/deployment.yaml#L178

dfengshi commented:
@bonifaido Hi Nandor, I've added the method `withStatsDVolume()` with the configmap setup as requested, but I'm fuzzy about what's going on in `deploymentForConfigurer()` and `statefulSetForVault()` vs what's going on in the helm chart. Should I be calling `withStatsDVolume()` in both of those methods? Maybe something like this: `withStatsDVolume(v, withTLSVolume(v, withCredentialsVolume(v, []v1.Volume{` ? 

This might be a bad question, because I might be missing some things in terms of understanding.

I've also added you as a collaborator on my fork in case it might be easier for you to just go ahead and dive in to my broken code rather than wait for me to understand what's missing.

Thanks again for the guidance+patience!

bonifaido commented:
Hi David, I have pushed a fix to your branch, if you don't mind I will rebase your commits into one, so two commits will arrive in the master branch. I have tested the operator and now it is working and exposing the Prometheus metrics correctly, thank you for your contribution! 👍

dfengshi commented:
@bonifaido Thanks Nandor!
